firedragon: init at 11.20.0-1

dr460nf1r3 opened the pull request:
## Description of changes

This PR adds `firedragon` to Nixpkgs. It is a fork of the Floorp browser, with custom branding and opinionated settings.

Source: [https://gitlab.com/garuda-linux/firedragon/settings](https://gitlab.com/garuda-linux/firedragon/settings)

## Todo

- [x] Get it to build (current error: `wasm-ld: error: unknown argument: --undefined-version`), [full build logs](https://bin.garudalinux.org/?f17768a1cc4e5db2#3V2sMTPhq5tCMtHfbsPrtPhrog6HtdftStaDnPU7JrTA) - this error originated in using `pkgs/by-name/fi/firedragon-unwrapped` as way to build `firedragon-unwrapped` (CI complained about using `callPackage` in `all-packages.nix`, doing it the "traditional" way gets rid of the error. I don't understand this enough to make it work the other way - does anyone else know whats going on here?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dr460nf1r3 committed:
firedragon: init at 11.20.0-1

dr460nf1r3 commented:
~~~
- Attribute `pkgs.firedragon-unwrapped` is a new top-level package using `pkgs.callPackage ./pkgs/applications/networking/browsers/firedragon { /* ... */ }`.
  Please define it in pkgs/by-name/fi/firedragon-unwrapped/package.nix instead.
  See `pkgs/by-name/README.md` for more details.
  Since the second `callPackage` argument is `{ }`, no manual `callPackage` in pkgs/top-level/all-packages.nix is needed anymore.

This PR introduces additional instances of discouraged patterns as listed above. Merging is discouraged but would not break the base branch.
To run locally: ./maintainers/scripts/check-by-name.sh master https://github.com/NixOS/nixpkgs.git
~~~

Doing it as outlined makes the entire build fail horribly like that: [log](https://bin.garudalinux.org/?f17768a1cc4e5db2#3V2sMTPhq5tCMtHfbsPrtPhrog6HtdftStaDnPU7JrTA). I

JohnRTitor reviewed (commented):
CCing @mweinelt, Firefox maintainer, to review this. (I am not sure if we need another firefox fork)

mweinelt commented:
I'm not a fan of adding more browser forks. They are not exactly cheap to build and likely don't attract a large enough user base to justify building them every other day.

caniko commented:
I really want this.

liberodark commented:
Hi,

i hope you can finish your PR @dr460nf1r3
And i close my pr for add firedragon because you are the best maintainer for this app.

Best Regards

caniko commented:
> I'm not a fan of adding more browser forks. They are not exactly cheap to build and likely don't attract a large enough user base to justify building them every other day.

So the line in the sand are forked projects? Horrible precedent.

dr460nf1r3 commented:
> Hi,
> 
> i hope you can finish your PR @dr460nf1r3 And i close my pr for add firedragon because you are the best maintainer for this app.
> 
> Best Regards

I would love to, but I didn't update it because it seemed like people were not interested in adding it :/

liberodark commented:
I have several users who use your application, that's why I started a PR.
Also that close this issue : https://github.com/NixOS/nixpkgs/issues/183035

dr460nf1r3 commented:
I updated the PR with the current version and added some needed things such as extra policy/prefs. That being said, I don't know have to fix the `nixpkgs-vet` check due to above mentioned issues.

JohnRTitor commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29218,6 +29218,13 @@ with pkgs;
     libName = "librewolf";
   };
 
+  firedragon-unwrapped = callPackage ../applications/networking/browsers/firedragon {};
```
To fix the CI you need to do exactly what it says, move to by-name as firedragon-unwrapped, then remove the above line from here. 

The firedragon override below should stay though.

dr460nf1r3 replied:
Well I was able to read that message 😋 but as stated in the PR I got an error doing it that way (`wasm-ld: error: unknown argument: --undefined-version`)

Aleksanaa replied:
> Well I was able to read that message 😋 but as stated in the PR I got an error doing it that way (`wasm-ld: error: unknown argument: --undefined-version`)

I'm pretty sure this is not related. Unless you have made your own (you must know) tricky patterns such as relative paths outside the import path.

jopejoe1 commented on pkgs/applications/networking/browsers/firedragon/default.nix:
```diff
@@ -0,0 +1,115 @@
+      inherit (nixosTests) firedragon;
+    };
+
+    extraConfigureFlags = [
+      "--disable-crashreporter"
+      "--disable-debug"
+      "--disable-debug-js-modules"
+      "--disable-debug-symbols"
+      "--disable-default-browser-agent"
+      "--disable-gpsd"
+      "--disable-necko-wifi"
+      "--disable-rust-tests"
+      "--disable-tests"
+      "--disable-updater"
+      "--disable-warnings-as-errors"
+      "--disable-webspeech"
+      "--enable-bundled-fonts"
+      "--enable-jxl"
+      "--enable-private-components"
+      "--enable-proxy-bypass-protection"
+      "--with-app-name=${pname}"
+      "--with-app-basename=${applicationName}"
+      "--with-distribution-id=org.garudalinux"
+      "--with-unsigned-addon-scopes=app,system"
+    ];
```
The following are already handled in `buildMozillaMach`:
- `debug` is already handled by the `debugBuild` flag `false` by default
- `necko-wifi` is already handled by the `crashreporterSupport` option;
- `crashreporter` is handled by `geolocationSupport` which is false when `privacySupport` is true;
- `tests` and `updater` are currently always disabled in `buildMozillaMach`
- `distribution-id` get set to `org.nixos` in `buildMozillaMach` and should stay like that even in forks of Firefox, as it identifies to the users that this build was distributed through nixpkgs.

```suggestion
    extraConfigureFlags = [
      "--disable-debug-js-modules"
      "--disable-debug-symbols"
      "--disable-default-browser-agent"
      "--disable-gpsd"
      "--disable-rust-tests"
      "--disable-warnings-as-errors"
      "--disable-webspeech"
      "--enable-bundled-fonts"
      "--enable-jxl"
      "--enable-private-components"
      "--enable-proxy-bypass-protection"
      "--with-app-name=${pname}"
      "--with-app-basename=${applicationName}"
      "--with-unsigned-addon-scopes=app,system"
    ];
```

dr460nf1r3 replied:
Thanks, fixed! :)

Aleksanaa commented on pkgs/applications/networking/browsers/firedragon/default.nix:
```diff
@@ -0,0 +1,109 @@
+    # Upstream already includes some of the bugfix patches that are applied by
+    # `buildMozillaMach`. Pick out only the relevant ones for Floorp and override
+    # the list here.
+    patches = [
+      ../firefox/env_var_for_system_dir-ff111.patch
+      ../firefox/no-buildconfig-ffx121.patch
+    ];
```
This reference is illegal.

A. Copy them here.
B. Add them to firefox's passthru and reference them here.

Aleksanaa commented:
I've checked that `firedragon-unwrapped.outPath` remains the same (after removing the patches) whether putting inside or outside of `by-name`. Your build error can only be caused by something else, e.g. you rebased to latest master, race condition in build scripts, etc

dr460nf1r3 commented:
I'll give it another try once I'm able to work on it. Thanks for verifying!
